src/mbi/relaxed_projection.py

```diff
@@ -0,0 +1,402 @@
+            domain: a Domain class.
+        """
+        self.D_prime = D
+        self.domain = domain
+    
+    @property 
+    def feature_indices_map(self):
+        map_dict = {}
+        s = 0
+        for col in self.domain.attrs:
+            e = s + self.domain.size(col)
+            map_dict[col] = list(range(s, e))
+            s = e
+        return map_dict
+
+    def datavector(self):
+        """ This will return a flattened vector of PROBABILITY distribution """
+        splits = [self.D_prime[:, self.feature_indices_map[attr]] for attr in self.domain.attrs]
+
+        letters = "bcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        assert(
+            len(letters) >= len(self.domain)
+        ), 'High dim domain'
+
+        input_dims = ','.join(f'a{letters[i]}' for i in range(len(splits)))
+        output_dims = ''.join(f'{letters[i]}' for i in range(len(splits)))
+        einsum_str = f'{input_dims}->a{output_dims}'
+
+        joint = jnp.einsum(einsum_str, *splits).mean(axis=0)
+        return joint.flatten()
+
+
+    def datamatrix(self):
+        """ This will return a matrix of frequency COUNT """
+        splits = [self.D_prime[:, self.feature_indices_map[attr]] for attr in self.domain.attrs]
+
+        letters = "bcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        assert(
+            len(letters) >= len(self.domain)
+        ), 'High dim domain'
+
+        input_dims = ','.join(f'a{letters[i]}' for i in range(len(splits)))
+        output_dims = ''.join(f'{letters[i]}' for i in range(len(splits)))
+        einsum_str = f'{input_dims}->a{output_dims}'
+
+        joint = jnp.einsum(einsum_str, *splits).sum(axis=0)
+        return joint
+
+
+    def project(self, cl: tuple):
+        cols = list(itertools.chain.from_iterable(self.feature_indices_map[attr] for attr in cl))
+        D_proj = self.D_prime[:, cols]
+        domain_proj = self.domain.project(cl)
+        return ProjectableData(D_proj, domain_proj)
+    
+
+    def synthetic_data(self, num_samples=1000, seed=0):
+        n, _ = self.D_prime.shape
+        key = jax.random.PRNGKey(seed)
+        key, sk = jax.random.split(key)
+        sampled_indices = jax.random.choice(key, n, shape=(num_samples,), replace=True)
+        sampled_data = self.D_prime[sampled_indices]
+
+        result = []
+        for attr in self.domain.attrs:
+            p = sampled_data[:, self.feature_indices_map[attr]]
+            key, sk = jax.random.split(key)
+            ids = jax.random.categorical(sk, jnp.log(jnp.clip(p, a_min=1e-30, a_max=1.0)), axis=-1)
+            result.append(ids)
+
+        sampled_df = pd.DataFrame(np.stack([np.array(r) for r in result], axis=1), columns = list(self.domain.attrs))
+        sampled_data = Dataset(sampled_df, self.domain)
+        return sampled_data
+
+
+def relaxed_projection_estimation(
+    domain: Domain,
+    loss_fn: marginal_loss.MarginalLossFn | list[LinearMeasurement],
+    iters: int = 1000,
+    **kwargs
```

Comment:
It's okay (and probably better) for these to be written out explicitly as you had before (even though the Protocol writes it as **kwargs)